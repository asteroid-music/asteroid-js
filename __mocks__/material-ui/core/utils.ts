function createSvgIcon(path,displayName) {
    return "Memo("+displayName+"Icon)";
}

export { createSvgIcon };
