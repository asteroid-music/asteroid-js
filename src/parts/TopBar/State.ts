/**
 * State interface for the <TopBar> component
 * Details for each individual property is given in comments below.
 */
interface TopBarState {
    /** Whether the sidebar should be open */
    sidebarOpen: boolean,
    /** Name of the currently open sidebar */
    sidebarName: string,
    /** Subtabs of the currently open sidebar */
    sidebarSubtabs: string[]
}

export default TopBarState;
