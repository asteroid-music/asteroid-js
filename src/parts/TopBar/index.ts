//Main component
export { default } from './TopBar';

//Props interface
export { default as TopBarProps } from './Props'

