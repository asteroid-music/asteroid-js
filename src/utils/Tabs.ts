/**
 * Interface describing an object with the same elements as SubTabObject
 */
interface SubTabObjectLike {
    /** The name of the tab */
    name: string;

    /** A string array containing names of subtabs */
    component: React.ComponentType<any>;
}

/**
 * Interface describing an object with the same elements as TabObject
 */
interface TabObjectLike {
    /** The name of the tab */
    name: string;

    /** A subtab array containing names of subtabs */
    subtabs: SubTabObjectLike[];
}

class SubTabObject {
    /** The name of the subtab */
    name: string;

    /** The component of the subtab */
    component: React.ComponentType<any>;

    /**
     * Constructor method
     *
     * @param {object} rawJson: a raw json representing the object
     */
    constructor(rawJson: SubTabObjectLike) {
        this.name = rawJson.name;
        this.component = rawJson.component;
    }

}

/**
 * Class describing named tab, along with strings representing all the subtabs
 * it contains
 */
class TabObject {
    /** The name of the tab */
    name: string;

    /** A string array containing names of subtabs */
    subtabs: SubTabObject[];

    /**
     * Constructor method
     *
     * @param {object} rawJson: a raw json representing the object
     */
    constructor(rawJson: TabObjectLike) {
        this.name = rawJson.name;
        this.subtabs = rawJson.subtabs.map((item: SubTabObjectLike) => {
            return new SubTabObject(item);
        });
    }

    /**
     * Gets a list of names of subtabs in the tab
     *
     * @returns {string[]} tablist: the name of each subtab
     */
    nameList() {
       return this.subtabs.map((subtab: SubTabObject) => {
            return subtab.name;
       });
    }

    /**
     * Checks if the subtab is present in the tab
     *
     * @param {string} subtab: the name of the subtab to check for
     *
     * @returns {boolean} isin: true if subtab is in the tab; false otherwise
     */
    includes(subtab:string) {
        return this.nameList().includes(subtab);
    }

    /**
     * Returns the subtab specified by the given subtab name,
     *      or 'null' if none exists
     *
     * @param {string} subtab: the name of the subtab to get
     *
     * @returns {SubTabObject | null} subtab: the specified subtab if exists;
     *      'null' otherwise
     */
    get(subtab:string) {
        let objIndex: number = this.nameList().indexOf(subtab);
        if (objIndex === -1) {
            return null;
        } else {
            return this.subtabs[objIndex];
        }
    }
}

/**
 * Class describing an array of named tabs
 */
class TabArray {
    /** Encapsulated array */
    tabs: TabObject[];

    /**
     * Constructor method
     *
     * @param {object} rawJson: a raw json representing the object
     */
    constructor(rawJson: TabObjectLike[]) {
        this.tabs = rawJson.map((item: TabObjectLike) => {
            return new TabObject(item);
        });
    }

    /**
     * Gets a list of names of tabs in the array
     *
     * @returns {string[]} tablist: the name of each tab
     */
    nameList() {
       return this.tabs.map((tab: TabObject) => {
            return tab.name;
       });
    }

    /**
     * Checks if the tab is present in the array
     *
     * @param {string} tab: the name of the tab to check for
     *
     * @returns {boolean} isin: true if tab is in the tab array;
     *      false otherwise
     */
    includes(tab:string) {
        return this.nameList().includes(tab);
    }

    /**
     * Returns the tab specified by the given tabname, or 'null' if none exists
     *
     * @param {string} tab: the name of the tab to get
     *
     * @returns {TabObject | null} tab: the specified tab if exists;
     *      'null' otherwise
     */
    get(tab:string) {
        let objIndex: number = this.nameList().indexOf(tab);
        if (objIndex === -1) {
            return null;
        } else {
            return this.tabs[objIndex];
        }
    }
}

export { TabObjectLike, TabObject, TabArray }
