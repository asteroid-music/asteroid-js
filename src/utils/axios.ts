import axios from 'axios';

//console.log(asteroidConfig);

declare var __config__: {
    "axios": {
        "baseURL": string,
        "timeout": number
    }
};

let instance = axios.create(__config__.axios || {});

export default instance;
