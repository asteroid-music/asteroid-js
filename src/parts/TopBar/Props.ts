//Imports from React module
import React from 'react';

//Import from local 'utils'
import { TabArray } from 'utils/Tabs';

/**
 * Interface for props for the <TopBar> React component.
 * Details for each individual property is given in comments below.
 */
interface TopBarProps {
    /** Props for react router */
    match: object,
    history: object,
    location: {hash: string, pathname: string, search: string},
    /** A TabArray object representing the accessible tabs */
    tabs: TabArray;
}

export default TopBarProps
