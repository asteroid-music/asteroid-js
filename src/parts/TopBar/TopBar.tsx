//Import from external module 'react'
import React from 'react';

//Imports from external module '@material-ui/core'
import Tabs from '@material-ui/core/Tabs';
import AppBar from '@material-ui/core/AppBar';
import Tab from '@material-ui/core/Tab';
import Paper from '@material-ui/core/Paper';
import Drawer from '@material-ui/core/Drawer';

//Imports from external module 'react-router'
import { withRouter } from 'react-router-dom';

//Imports from local folder
import TopBarProps from './Props';
import TopBarState from './State';

//Imports from local 'parts'
import SideBar from 'parts/SideBar';

/**
 * <TopBar> React component for navigating Asteroid.
 * Creates a bar at the top of the screen with buttons to access tabs and
 *      subtabs
 */
class TopBar extends React.Component<TopBarProps,TopBarState> {
    constructor(props) {
        super(props);
        this.state = {
            sidebarOpen: false,
            sidebarName: "",
            sidebarSubtabs: []
        };
    }

    render() {
        let location: string | undefined = this.props.location.pathname.split("/")[1];
        return (
            <AppBar position="static">
                {/*Main tab bar*/}
                <Tabs
                    value={location || false}
                    onChange={
                        (event: React.ChangeEvent<{}>, value: string) => {
                            this.setState({
                                sidebarOpen: true,
                                sidebarName: value,
                                sidebarSubtabs: this.props.tabs.get(value)?.nameList() || []
                            });
                        }
                    }
                    variant="scrollable"
                    scrollButtons="on"
                >
                    {this.props.tabs.nameList().map((tabName: string) => {
                        return <Tab
                            key={tabName}
                            value={tabName}
                            label={tabName}
                        />
                    })}
                </Tabs>
                <SideBar
                    open={this.state.sidebarOpen}
                    tabName={this.state.sidebarName}
                    linkNames={this.state.sidebarSubtabs}
                    closeCallback={()=>{this.setState({sidebarOpen: false});}}
                />
            </AppBar>
        );
    }
}

export default withRouter(TopBar);
