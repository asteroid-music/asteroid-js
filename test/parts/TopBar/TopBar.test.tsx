import React from 'react';
import renderer from 'react-test-renderer';
import TopBar from '../../../src/parts/TopBar';
import { serialize as serialise } from 'jest-snapshot/build/utils';
//I know, UK vs US spelling...

const testTabNames = [
    "Test tab name 1",
    "Test tab name 2",
    "Test tab name 3"
]

const testSubTabNames = [
    "Test subtab name 1",
    "Test subtab name 2",
    "Test subtab name 3"
]

/**
 * A hacky work-around to unmounting the element.
 * Because material-ui's <Tabs /> uses a forwardRef that
 * then is utilised in unmounting it, but react-test-renderer
 * can't handle refs, unmounting will throw up an error from the
 * <Tabs /> component. However, if this error is caught and ignored,
 * the unmount will still have been successful, allowing further
 * creation of components using renderer.create()
 */
function hackUnmount(tree) {
    jest.spyOn(console, 'error').mockImplementation(() => {});
    try {
        tree.unmount();
    } catch(error) {}
}

/**
 * As MUI uses clsx to procedurally generate class names,
 * there will be near matches which differ only by the number of class.
 * For some checks, these should be ignored
 */
function replaceClassNumbers(serialised:string) {
    const numRegex = /-\d+/;
    return serialised.split("\n").map(line=>{
        if (line.includes("className")) {
            while (line.match(numRegex)) {
                line = line.replace(numRegex,"-Number");
            }
        }
        return line;
    }).join("\n");
}

describe('<TopBar />',()=>{
    //Basic rendering with minimal props
    it("renders correctly",()=>{
        let tree = renderer.create(<TopBar
                tabNames={testTabNames}
                tabCallback={()=>{}}
                subTabCallback={()=>{}}
            />
        )
        const json = tree.toJSON();
        hackUnmount(tree);
        expect(json).toMatchSnapshot();
    });

    //Basic rendering with minimal props
    it("successfully calls tabCallback",()=>{
        const tabCallback = jest.fn();
        let tree = renderer.create(<TopBar
                tabNames={testTabNames}
                tabCallback={()=>{tabCallback();}}
                subTabCallback={()=>{}}
            />
        )
        tree.root.findAllByType('button').forEach((button)=>{
            button.props.onClick();
        })
        hackUnmount(tree);
        expect(tabCallback).toHaveBeenCalledTimes(testTabNames.length);
    });

    //Rendering with currTab prop
    describe('prop: currTab',()=>{
        it("renders correctly with valid tab name",()=>{
            let tree = renderer
                .create(<TopBar
                    tabNames={testTabNames}
                    tabCallback={()=>{}}
                    subTabCallback={()=>{}}
                    currTab={testTabNames[1]}
                />);
            const json = tree.toJSON();
            hackUnmount(tree);
            expect(json).toMatchSnapshot();
        });

        it("ignores an invalid tab name",()=>{
            let tree = renderer
                .create(<TopBar
                    tabNames={testTabNames}
                    tabCallback={()=>{}}
                    subTabCallback={()=>{}}
                    currTab={"this is not a valid tab name"}
                />);
            const firstString = serialise(tree.toJSON());
            hackUnmount(tree);
            tree = renderer.create(
                <TopBar
                    tabNames={testTabNames}
                    tabCallback={()=>{}}
                    subTabCallback={()=>{}}
                />
            );
            const secondString = serialise(tree.toJSON());
            hackUnmount(tree);
            expect(
                secondString
            ).toEqual(
                firstString
            );
        });
    });

    //Rendering with subTabNames prop
    describe('prop: subTabNames',()=>{
        it("renders with a subtab bar",()=>{
            let tree = renderer
                .create(<TopBar
                    tabNames={testTabNames}
                    subTabNames={testSubTabNames}
                    tabCallback={()=>{}}
                    subTabCallback={()=>{}}
                />);
            const json = tree.toJSON();
            hackUnmount(tree);
            expect(json).toMatchSnapshot();
        });
    });


    //Rendering with currSubTab prop
    describe('prop: currTab',()=>{
        it("renders correctly with valid tab and subtab name",()=>{
            let tree = renderer
                .create(<TopBar
                    tabNames={testTabNames}
                    subTabNames={testSubTabNames}
                    tabCallback={()=>{}}
                    subTabCallback={()=>{}}
                    currTab={testTabNames[1]}
                    currSubTab={testSubTabNames[1]}
                />);
            const json = tree.toJSON();
            hackUnmount(tree);
            expect(json).toMatchSnapshot();
        });

        it("ignores an invalid subtab name with valid tab",()=>{
            let tree = renderer
                .create(<TopBar
                    tabNames={testTabNames}
                    subTabNames={testSubTabNames}
                    tabCallback={()=>{}}
                    subTabCallback={()=>{}}
                    currTab={testTabNames[1]}
                />);
            const firstString = replaceClassNumbers(serialise(tree.toJSON()));
            hackUnmount(tree);
            tree = renderer
                .create(<TopBar
                    tabNames={testTabNames}
                    subTabNames={testSubTabNames}
                    tabCallback={()=>{}}
                    subTabCallback={()=>{}}
                    currTab={testTabNames[1]}
                    currSubTab={"this is not a valid subtab name"}
                />);
            const secondString = replaceClassNumbers(serialise(tree.toJSON()));
            hackUnmount(tree);
            expect(
                secondString
            ).toEqual(
                firstString
            );
        });

        it("ignores a valid subtab name with no subTabNames",()=>{
            let tree = renderer
                .create(<TopBar
                    tabNames={testTabNames}
                    tabCallback={()=>{}}
                    subTabCallback={()=>{}}
                    currTab={testTabNames[1]}
                    currSubTab={testSubTabNames[1]}
                />);
            const firstString = replaceClassNumbers(serialise(tree.toJSON()));
            hackUnmount(tree);
            tree = renderer
                .create(<TopBar
                    tabNames={testTabNames}
                    tabCallback={()=>{}}
                    subTabCallback={()=>{}}
                    currTab={testTabNames[1]}
                />);
            const secondString = replaceClassNumbers(serialise(tree.toJSON()));
            hackUnmount(tree);
            expect(
                secondString
            ).toEqual(
                firstString
            );
        });
    });
});
