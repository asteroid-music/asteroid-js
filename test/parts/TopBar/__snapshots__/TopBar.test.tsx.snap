// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopBar /> renders correctly 1`] = `
<header
  className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
>
  <div
    className="MuiTabs-root"
  >
    <div
      aria-disabled={false}
      className="MuiButtonBase-root MuiTabScrollButton-root MuiTabs-scrollButtons Mui-disabled"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role={null}
      tabIndex={null}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
        />
      </svg>
    </div>
    <div
      className="MuiTabs-scrollable"
      style={
        Object {
          "height": 99,
          "overflow": "scroll",
          "position": "absolute",
          "top": -9999,
          "width": 99,
        }
      }
    />
    <div
      className="MuiTabs-scroller MuiTabs-scrollable"
      onScroll={[Function]}
      style={
        Object {
          "marginBottom": null,
          "overflow": "hidden",
        }
      }
    >
      <div
        className="MuiTabs-flexContainer"
        onKeyDown={[Function]}
        role="tablist"
      >
        <button
          aria-selected={false}
          className="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="tab"
          tabIndex={-1}
          type="button"
        >
          <span
            className="MuiTab-wrapper"
          >
            TestTab
          </span>
        </button>
      </div>
    </div>
    <div
      aria-disabled={false}
      className="MuiButtonBase-root MuiTabScrollButton-root MuiTabs-scrollButtons Mui-disabled"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role={null}
      tabIndex={null}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
        />
      </svg>
    </div>
  </div>
</header>
`;
